src/audio/mux/mux_generic.c

```diff
@@ -18,347 +18,528 @@
 #include <stddef.h>
 #include <stdint.h>
 
-#if CONFIG_FORMAT_S16LE
-/*
- * \brief Fetch 16b samples from source buffer and perform routing operations
- *	  based on mask provided.
- * \param[in,out] source Source buffer.
- * \param[in] offset Offset in source buffer.
- * \param[in] mask Routing bitmask for calculating output sample.
- */
-UT_STATIC inline int32_t calc_sample_s16le(const struct audio_stream *source,
-					   uint32_t offset, uint8_t mask)
+static void mux_check_for_wrap(struct audio_stream *sink,
+			       const struct audio_stream **sources,
+			       struct mux_look_up *lookup)
 {
-	int32_t sample = 0;
-	int16_t *src;
-	int8_t in_ch;
+	const struct audio_stream *source;
+	uint32_t elem;
+
+	/* check sources and destinations for wrap */
+	for (elem = 0; elem < lookup->num_elems; elem++) {
+		source = sources[lookup->copy_elem[elem].stream_id];
+		lookup->copy_elem[elem].dest =
+			audio_stream_wrap(sink, lookup->copy_elem[elem].dest);
+		lookup->copy_elem[elem].src =
+			audio_stream_wrap(source, lookup->copy_elem[elem].src);
+	}
+}
 
-	if (mask == 0)
-		return 0;
+static void demux_check_for_wrap(struct audio_stream *sink,
+				 const struct audio_stream *source,
+				 struct mux_look_up *lookup)
+{
+	uint32_t elem;
+
+	/* check sources and destinations for wrap */
+	for (elem = 0; elem < lookup->num_elems; elem++) {
+		lookup->copy_elem[elem].dest =
+			audio_stream_wrap(sink, lookup->copy_elem[elem].dest);
+		lookup->copy_elem[elem].src =
+			audio_stream_wrap(source, lookup->copy_elem[elem].src);
+	}
+}
 
-	for (in_ch = 0; in_ch < source->channels; in_ch++) {
-		if (mask & BIT(in_ch)) {
-			src = audio_stream_read_frag_s16(source,
-							 offset + in_ch);
-			sample += *src;
-		}
+static uint32_t mux_calc_frames_without_wrap_s32(struct comp_dev *dev,
+						 struct audio_stream *sink,
+						 const struct audio_stream
+						 **sources,
+						 struct mux_look_up *lookup)
+{
+	const struct audio_stream *source;
+	uint32_t frames;
+	uint32_t min_frames;
+	uint32_t elem;
+	void *ptr;
+
+	/* dest pointer for all copy_elems in lookup refers to the same
+	 * sink buffer (mux has one sink buffer), so dest min_frames
+	 * calculation based only on lookup table first element is sufficient.
+	 */
+	ptr = (int32_t *)lookup->copy_elem[0].dest -
+		lookup->copy_elem[0].out_ch;
+	min_frames = audio_stream_frames_without_wrap(sink, ptr);
+
+	for (elem = 0; elem < lookup->num_elems; elem++) {
+		source = sources[lookup->copy_elem[elem].stream_id];
+
+		ptr = (int32_t *)lookup->copy_elem[0].src -
+			lookup->copy_elem[0].in_ch;
```

Comment:
Btw, why is copy_elem[0] hard coded here, should this be the array "elem" index ?

Reply:
My fault, updated.